File: Algorithms/cpp/024.SwapNodesInPairs.cpp
class Solution {
public:
    ListNode *swapPairs(ListNode *head) {
        ListNode *dummyHead = new ListNode(1);
        dummyHead->next = head;
        ListNode *prev = dummyHead;
        ListNode *p = head;
        while (p != NULL && p->next != NULL) {
            ListNode *q = p->next;
            ListNode *r = p->next->next;
            prev->next = q;
            q->next = p;
            p->next = r;
            prev= p;
            p = r;
        }
        return dummyHead->next;
    }
};

File: Algorithms/py/024.SwapNodesInPairs.py
class Solution:
    # @param a ListNode
    # @return a ListNode
    def swapPairs(self, head):
        dummy= ListNode(1)
        dummy.next = head
        prev, cur = dummy, head
        while cur and cur.next:
            q, r = cur.next, cur.next.next
            prev.next, q.next, cur.next = q, cur, r
            prev, cur = cur, r
        return dummy.next
